sdks/java/io/jdbc/src/main/java/org/apache/beam/sdk/io/jdbc/JdbcIO.java

```diff
@@ -635,6 +596,131 @@ public void teardown() throws Exception {
+     */
+    public Write<T> withBatchSize(long batchSize) {
+      checkArgument(batchSize > 0, "batchSize must be > 0, but was %d", batchSize);
+      return toBuilder().setBatchSize(batchSize).build();
+    }
+
+    @Override
+    protected WriteFn<T> createWriteFn(AbstractWrite<T, T> write) {
+      return new WriteFn<T>(write);
+    }
+
+    private static class WriteFn<T> extends AbstractWriteFn<T, T> {
+
+      public WriteFn(AbstractWrite<T, T> spec) {
+        super(spec);
+      }
+
+      @ProcessElement
+      @Override
+      public void processElement(ProcessContext context) throws Exception {
+        T record = context.element();
+
+        preparedStatement.clearParameters();
+        spec.getPreparedStatementSetter().setParameters(record, preparedStatement);
+        preparedStatement.addBatch();
+
+        batchCount++;
+
+        if (batchCount >= spec.getBatchSize()) {
+          executeBatch();
+        }
+      }
+    }
+  }
+
+  /**
+   * A {@link PTransform} to write to a JDBC datasource where the input is an
+   * Iterable of the row type.
+   */
+  @AutoValue
+  public abstract static class WriteIterable<T> extends AbstractWrite<T, Iterable<T>> {
+    // Override is needed, otherwise the code generator complains about the type parameter
+    @Nullable abstract PreparedStatementSetter<T> getPreparedStatementSetter();
+
+    abstract Builder<T> toBuilder();
+
+    @AutoValue.Builder
+    abstract static class Builder<T> {
+      abstract Builder<T> setDataSourceConfiguration(DataSourceConfiguration config);
+      abstract Builder<T> setStatement(String statement);
+      abstract Builder<T> setPreparedStatementSetter(PreparedStatementSetter<T> setter);
+
+      abstract WriteIterable<T> build();
+    }
+
+    public WriteIterable<T> withDataSourceConfiguration(DataSourceConfiguration config) {
+      return toBuilder().setDataSourceConfiguration(config).build();
+    }
+    public WriteIterable<T> withStatement(String statement) {
+      return toBuilder().setStatement(statement).build();
+    }
+    public WriteIterable<T> withPreparedStatementSetter(PreparedStatementSetter<T> setter) {
+      return toBuilder().setPreparedStatementSetter(setter).build();
+    }
+
+    @Override
+    protected WriteIterableFn<T> createWriteFn(AbstractWrite<T, Iterable<T>> write) {
+      return new WriteIterableFn<T>(write);
+    }
+
+    private static class WriteIterableFn<T> extends AbstractWriteFn<T, Iterable<T>> {
+
+      public WriteIterableFn(AbstractWrite<T, Iterable<T>> spec) {
+        super(spec);
+      }
+
+      @ProcessElement
+      @Override
+      public void processElement(ProcessContext context) throws Exception {
+        Iterable<T> records = context.element();
+
+        for (T record : records) {
+          preparedStatement.clearParameters();
+          spec.getPreparedStatementSetter().setParameters(record, preparedStatement);
+          preparedStatement.addBatch();
```

Comment:
What if the Iterable has too many elements for the database to handle in a single batch prepared statement?